// @clerk/types/dist/appearance.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
exports.__esModule = true;
exports.shadesOfPurple = exports.socialButtonsRoundedPills = exports.dark = exports.createClerkTheme = void 0;
// TODO: Remove before GA
var createClerkTheme = function (appearance) {
    // Placeholder method that might hande more transformations in the future
    return __assign(__assign({}, appearance), { __type: 'prebuilt_appearance' });
};
exports.createClerkTheme = createClerkTheme;
exports.dark = (0, exports.createClerkTheme)({
    variables: {
        colorBackground: 'rgba(43,43,56)',
        colorInputBackground: 'rgba(255,255,255,0.1)',
        colorAlphaShade: 'white',
        colorText: 'white',
        colorInputText: 'white'
    },
    elements: function () {
        return {
            socialButtonsLogo__apple: { filter: 'invert(1)' },
            socialButtonsLogo__github: { filter: 'invert(1)' },
            activeDeviceIcon: {
                '--cl-chassis-bottom': '#d2d2d2',
                '--cl-chassis-back': '#e6e6e6',
                '--cl-chassis-screen': '#e6e6e6',
                '--cl-screen': '#111111'
            }
        };
    }
});
exports.socialButtonsRoundedPills = (0, exports.createClerkTheme)({
    layout: {
        socialButtonsVariant: 'iconButton'
    },
    elements: function (_a) {
        var theme = _a.theme;
        return ({
            socialButtons: {
                gridTemplateColumns: "repeat(auto-fit, ".concat(theme.space.$12, ")")
            },
            socialButtonsButtonIcon: {
                borderColor: theme.colors.$primary500,
                borderRadius: '999px'
            }
        });
    }
});
exports.shadesOfPurple = (0, exports.createClerkTheme)({
    baseTheme: [exports.dark],
    variables: {
        colorBackground: '#3f3c77',
        colorPrimary: '#f8d80d',
        colorTextOnPrimaryBackground: '#38375f',
        colorInputText: '#a1fdfe'
    }
});
